### PR TITLE
Update storage format specification for Mbed TLS 2.25.0+

### PR DESCRIPTION
**Pull request by gilles-peskine-arm**
I haven't updated the historical information. That would be nice to do, but I ran out of time collecting information.

Fulfils https://github.com/Mbed-TLS/mbedtls/issues/3737 with respect to the current (and only supported) format.

Backport: https://github.com/Mbed-TLS/mbedtls/pull/5996

### CODE REVIEW

File: docs/architecture/mbed-crypto-storage-specification.md [hunk 1]
```diff
@@ -314,3 +314,134 @@ The layout of a key file is:
...
+
+The way in which the file name is constructed from the key identifier depends on the storage backend. The content of the file is described [below](#key-file-format-for-mbed-tls-2.25.0).
+
+* Library integration: the key file name is just the key identifier as defined in the PSA crypto specification. This is a 32-bit value which must be in the range 0x00000001..0x3fffffff (`PSA_KEY_ID_USER_MIN`..`PSA_KEY_ID_USER_MAX`).
+* PSA service integration: the key file name is `(uint32_t)owner_uid << 32 | key_id` where `key_id` is the key identifier from the owner point of view and `owner_uid` (of type `int32_t`) is the calling partition identifier provided to the server by the partition manager. This is a 64-bit value.
```
Comment (mpg) [inline docs/architecture/mbed-crypto-storage-specification.md:386]: `(uint32_t)owner_uid << 32 | key_id` does not look like correct C to me, I assume you mean `uint64_t` here?

**Review by mpg — COMMENTED** [review]
This is only a first pass, for now I have only checked that the spec is unambiguous and makes sense.

File: docs/architecture/mbed-crypto-storage-specification.md [hunk 2]
```diff
@@ -314,3 +314,134 @@ The layout of a key file is:
...
+
+* [Persistent keys](#key-file-format-for-mbed-tls-2.25.0) designated by a [key identifier and owner](#key-names-for-mbed-tls-2.25.0). Keys can be:
+    * Transparent, stored in the export format.
+    * Opaque, using the unified driver interface with statically registered drivers (`MBEDTLS_PSA_CRYPTO_DRIVERS`). The driver determines the content of the opaque key blob.
+    * Opaque, using the deprecated secure element interface with dynamically registered drivers (`MBEDTLS_PSA_CRYPTO_SE_C`). The driver picks a slot number which is stored in the place of the key material.
+* [Nonvolatile random seed](#nonvolatile-random-seed-file-format-for-mbed-tls-2.25.0) on ITS only.
+
+### Changes introduced in Mbed TLS 2.25.0
+
+* The numerical encodings of `psa_key_type_t`, `psa_key_usage_t` and `psa_algorithm_t` have changed.
```
Comment (mpg) [inline docs/architecture/mbed-crypto-storage-specification.md:345]: Also, it looks like the name of the undocumented macro for storage on stdio has changed from `CRYPTO_STORAGE_FILE_LOCATION` to `PSA_ITS_STORAGE_PREFIX`. Considering it's described as undocumented, perhaps the omission in the section was intentional, just mentioning it to be sure, not necessarily requesting a change here.

**Review by mpg — CHANGES_REQUESTED** [review]
I've completed my review to the depth the the time pressure allows. The only change that I'm really requesting is the use of `uint64_t` when shifting 32 places, the rest is informational or just a suggestion.

File: docs/architecture/mbed-crypto-storage-specification.md (hunk 2 above)
Comment (gilles-peskine-arm) [reply docs/architecture/mbed-crypto-storage-specification.md:345]: But actually the macro name changed at the same time as the format — but not in 2.25, a lot earlier than that. Update incoming.

**Comment by gilles-peskine-arm** [comment]
Regarding the accuracy of the documentation: when I wrote storage tests, I based my test code on the documentation, and the tests passed. But you have to take my word for that, and not all aspects of the format are tested.

**Review by mpg — APPROVED** [review]
Thanks for addressing my feedback, even the optional parts. Looks all good to me now!

**Pull request merged**